### PR TITLE
fix(tracestate): drop right-most member in tracestate

### PR DESCRIPTION
**Pull request by 1046102779**
Signed-off-by: 1046102779 <seachen@tencent.com>

https://github.com/w3c/trace-context/pull/448

### CODE REVIEW

File: trace/tracestate_test.go [hunk 1]
```diff
@@ -443,14 +443,21 @@ func TestTraceStateInsert(t *testing.T) {
...
-			name:       "too many entries",
+			name:       "drop the right-most member(oldest) in queue",
 			tracestate: maxMembers,
 			key:        "keyx",
 			value:      "valx",
-			expected:   maxMembers,
-			err:        errMemberNumber,
-		},
-	}
+			expected: func() TraceState {
```
Comment (dmathieu) [inline trace/tracestate_test.go:450]: I'm not finding this method super as to what it's doing. How about something like this:

```go
expected: func() TraceState {
	// Prepend the new element and remove the oldest one, which is over capacity.
	return TraceState{
		list: append(
			[]member{{Key: "keyx", Value: "valx"}},
			maxMembers.list[:len(maxMembers.list)-1]...,
		),
	}
}(),
```

**Review by dmathieu — COMMENTED** [review]
This will need a changelog entry

File: trace/tracestate.go [hunk 2]
```diff
@@ -179,17 +179,14 @@ func (ts TraceState) Insert(key, value string) (TraceState, error) {
 	}
```
Comment (MrAlias) [inline trace/tracestate.go:179]: The `Insert` function documentation needs to be updated to describe the new behavior. It should not mention an error is returned when capacity is exceeded.

File: trace/tracestate.go [hunk 3]
```diff
@@ -179,17 +179,14 @@ func (ts TraceState) Insert(key, value string) (TraceState, error) {
...
-		// https://github.com/w3c/trace-context/pull/448
-		return ts, fmt.Errorf("failed to insert: %w", errMemberNumber)
+	if cTS.Len()+1 <= maxListMembers {
+		cTS.list = append(cTS.list, member{})
 	}
-
-	cTS.list = append(cTS.list, member{})
+	// if tracestate's members > maxListMembers, defaultly
+	// it should drop the "right-most" member and insert the new
+	// member at the front.
```
Comment (MrAlias) [inline trace/tracestate.go:187]:
```suggestion
	// When the number of members exceeds capacity, drop the "right-most".
```

File: CHANGELOG.md [hunk 4]
```diff
@@ -17,6 +17,7 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 ### Changed
 
 - Jaeger exporter takes into additional 70 bytes overhead into consideration when sending UDP packets (#2489, #2512)
+- For tracestate's members, prepend the new element and remove the oldest one, which is over capacity (#2592)
```
Comment (MrAlias) [inline CHANGELOG.md:20]: Please move this to the a new `### Changed` sub-section of `## Unreleased`.

File: trace/tracestate.go (hunk 2 above)
Comment (MrAlias) [reply trace/tracestate.go:179]: This was resolved, but the original comment remains:

https://github.com/open-telemetry/opentelemetry-go/blob/d0068b24aa1a6162dc7efd306202a0f0f74ca8c5/trace/tracestate.go#L172-L174

Is there commit that didn't get pushed yet?

**Review by MrAlias — APPROVED** [review]

**Review by MadVikingGod — APPROVED** [review]

**Pull request merged**